src/Features/Core/Portable/NavigateTo/NavigateToProjectSearcher.cs

```diff
@@ -0,0 +1,347 @@
+                // the final result.
+                var documentSearchTasks = ArrayBuilder<Task>.GetInstance();
+                var tempArrays = ArrayBuilder<ArrayBuilder<INavigateToSearchResult>>.GetInstance();
+
+                try
+                {
+                    // Spawn off a task per document to search.
+                    SpawnDocumentSearchTasks(documentSearchTasks, tempArrays);
+
+                    // Wait for all the tasks to finish.
+                    await Task.WhenAll(documentSearchTasks).ConfigureAwait(false);
+                    _cancellationToken.ThrowIfCancellationRequested();
+
+                    // Collect all the results into the final result.
+                    var result = ArrayBuilder<INavigateToSearchResult>.GetInstance();
+                    foreach (var tempArray in tempArrays)
+                    {
+                        result.AddRange(tempArray);
+                    }
+
+                    return result.ToImmutableAndFree();
+                }
+                finally
+                {
+                    foreach (var tempArray in tempArrays)
+                    {
+                        tempArray.Free();
+                    }
+
+                    tempArrays.Free();
+                    documentSearchTasks.Free();
+                }
+            }
+
+            private void SpawnDocumentSearchTasks(ArrayBuilder<Task> tasks, ArrayBuilder<ArrayBuilder<INavigateToSearchResult>> tempArrays)
+            {
+                foreach (var document in _project.Documents)
+                {
+                    _cancellationToken.ThrowIfCancellationRequested();
+
+                    if (_searchDocument != null && document != _searchDocument)
+                    {
+                        continue;
+                    }
+
+                    var tempArray = ArrayBuilder<INavigateToSearchResult>.GetInstance();
+                    tempArrays.Add(tempArray);
+
+                    // Kick off a fresh task so searching can happen in parallel.
+                    tasks.Add(Task.Run(() => SearchDocumentAsync(document, tempArray)));
+                }
+            }
+
+            private async Task SearchDocumentAsync(
+                Document document, ArrayBuilder<INavigateToSearchResult> results)
+            {
+                _cancellationToken.ThrowIfCancellationRequested();
+
+                // Delay creating a semantic model until necessary.  But once we create it,
+                // cache for the remainder of the search so that it doesn't get cleaned up.
+                SemanticModel semanticModel = null;
+
+                var declarationInfo = await document.GetSyntaxTreeIndexAsync(_cancellationToken).ConfigureAwait(false);
+                foreach (var declaredSymbolInfo in declarationInfo.DeclaredSymbolInfos)
+                {
+                    if (_cancellationToken.IsCancellationRequested)
+                    {
+                        continue;
```

Comment:
:question: Why not return?

:question: Why return partial results for this cancellation, but in other cases we transition the task to a cancelled state?